pixeltable/catalog/catalog.py

```diff
@@ -1,32 +1,95 @@
 from __future__ import annotations
 
 import dataclasses
+import functools
 import logging
-from typing import Optional, Type
+import time
+from typing import TYPE_CHECKING, Any, Callable, Optional, TypeVar
 from uuid import UUID
 
+import psycopg
 import sqlalchemy as sql
 
-import pixeltable.env as env
 import pixeltable.exceptions as excs
 import pixeltable.metadata.schema as schema
 from pixeltable.env import Env
-
+from pixeltable.iterators import ComponentIterator
 from .dir import Dir
+from .globals import IfExistsParam, IfNotExistsParam, MediaValidation
+from .insertable_table import InsertableTable
+from .path import Path
 from .schema_object import SchemaObject
 from .table import Table
 from .table_version import TableVersion
 from .table_version_handle import TableVersionHandle
 from .table_version_path import TableVersionPath
+from .view import View
+
+if TYPE_CHECKING:
+    from .. import DataFrame, exprs
 
-# from .. import InsertableTable
 
 _logger = logging.getLogger('pixeltable')
 
 
-def _join_path(path: str, name: str) -> str:
-    """Append name to path, if path is not empty."""
-    return name if path == '' else f'{path}.{name}'
+def _lock_str(for_update: bool) -> str:
+    return 'X' if for_update else 'S'
+
+
+# TODO: remove once the concurrent update behavior has been debugged
+# def _debug_print(for_update: bool, msg: str) -> None:
+#     return
+#     print(f'{datetime.datetime.now()}: {_lock_str(for_update)}: {msg}')
+
+
+def _unpack_row(
+    row: Optional[sql.engine.Row], entities: list[type[sql.orm.decl_api.DeclarativeBase]]
+) -> Optional[list[Any]]:
+    """Convert a Row result into a list of entity instances.
+
+    Assumes that the query contains a select() of exactly those entities.
+    """
+    if row is None:
+        return None
+
+    result: list[sql.orm.decl_api.DeclarativeBase] = []
+    column_offset = 0
+
+    for entity in entities:
+        num_cols = len(entity.__table__.columns)
+        data = {name: row[column_offset + i] for i, name in enumerate(entity.__table__.columns.keys())}
+        inst = entity(**data)
+        result.append(inst)
+        column_offset += num_cols
+
+    return result
+
+
+_MAX_RETRIES = 3
+T = TypeVar('T')
+
+
+def _retry_loop(op: Callable[..., T]) -> Callable[..., T]:
+    @functools.wraps(op)
+    def loop(*args: Any, **kwargs: Any) -> T:
+        num_remaining_retries = _MAX_RETRIES
+        while True:
+            try:
+                # in order for retry to work, we need to make sure that there aren't any prior db updates
+                # that are part of an ongoing transaction
```

Comment:
Is this going to cause a problem with user-defined transactions? (i.e., what if a transaction block surrounds two user-level catalog operations?)

Reply:
They won't. User-level transactions will only be applicable to data operations.